Testing bounding boxes in WithMixedPrecisionTest

jaygala223 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons --sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #1784 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

Requesting @ianstenbit and @jbischof for review.

<!-- 
Feel free to tag @ianstenbit and @jbischof in your reviews.
-->

<!--
This PR template is copied and modified from here:
https://github.com/huggingface/transformers/blob/main/.github/PULL_REQUEST_TEMPLATE.md
-->

jaygala223 commented:
Hello @ianstenbit, I have added bounding boxes as inputs and made some changes to the test configs as well. Need to fix some failing tests as well. Is this a step in the right direction? Here's a screenshot of the tests for your reference:
![image](https://github.com/keras-team/keras-cv/assets/57001778/c7da75ca-92f0-4239-a25a-8376d653f4dd)

jaygala223 commented:
Thanks @ianstenbit, the output is pretty long. How can I share it with you? Here are a few screenshots:
![image](https://github.com/keras-team/keras-cv/assets/57001778/fb87000b-58c7-4326-b425-bdc291c61d96)
![image](https://github.com/keras-team/keras-cv/assets/57001778/a6776055-0c78-4c07-8c36-e2c7df1451f1)
![image](https://github.com/keras-team/keras-cv/assets/57001778/10faa496-fb61-40ae-8223-7e30fb007480)

ianstenbit commented:
> Here are a few screenshots

Thanks for sharing! In general, just the error message is a good place to start (in this case it's the `ValueError`)

I left a comment describing the issue (See above)
Thanks!

jaygala223 commented on keras_cv/layers/preprocessing/with_mixed_precision_test.py:
```diff
@@ -161,12 +164,26 @@ def test_can_run_in_mixed_precision(self, layer_cls, init_args):
-        inputs = {"images": img, "labels": labels}
+        
+        bounding_boxes = {
+            "boxes": tf.convert_to_tensor(
```
Got it. Thanks

jaygala223 commented:
Thanks for your help @ianstenbit. I did exactly as you suggested and we now stand at 19F, 10P, and 4 Skipped.

This keeps popping up: ``To fix, update the `fn_output_signature` (or `dtype`) argument to `map_fn`` along with some notImplementedErrors

Short summary of the remaining failures below:

![image](https://github.com/keras-team/keras-cv/assets/57001778/f368c64b-a7b6-48f6-a239-bf316f500f6b)
![image](https://github.com/keras-team/keras-cv/assets/57001778/e111a18d-662c-4cd1-b124-cfb3e2bd456f)

ianstenbit commented:
@jaygala223 some of these issues look like they may be because the layer under test doesn't support bounding boxes (like RandomZoom) -- in those cases we'll need to add box support to those layers.

I'll patch your CL and look at the other errors in some more detail

ianstenbit commented:
Turns out 5 of the issues were due to using `to_ragged` without a dtype specified.

I've fixed those on your dev branch, and I've also fixed one other one to give you a sense of what the changes look like to resolve the remaining issues.

ianstenbit commented:
There are 13 remaining failures -- the `ValueError` ones should be straightforward to sort out (like I did above for RandomCrop)

The `TypeError` I haven't looked at.

The `NotImplementedError` we can fix by adding a flag to the test (like the existing `NO_CPU_FP16_KERNEL_LAYERS` list) for layers that don't support boxes yet (or we can add box support!)

jaygala223 commented:
Thanks a lot, @ianstenbit. I'll add a flag (for now) to the tests raising a `NotImplementedError` and I'll also take a look at the other failures!

jaygala223 commented:
Also, one of the tests is now showing this. A StackOverflow [thread](https://stackoverflow.com/questions/51248813/any-ideas-on-what-might-be-causing-this-tensorflow-error-object-was-never-used) suggests that it may be an internal bug with tf v1.2 but I'm on 2.12🤔
![image](https://github.com/keras-team/keras-cv/assets/57001778/4f1a63c1-137d-436f-8035-47507b05c928)

ianstenbit commented:
> Also, one of the tests is now showing this. A StackOverflow [thread](https://stackoverflow.com/questions/51248813/any-ideas-on-what-might-be-causing-this-tensorflow-error-object-was-never-used) suggests that it may be an internal bug with tf v1.2 but I'm on 2.12🤔

This is indeed a very weird error message. Looking at the error logs, it seems that there's a dtype issue in the same test (RandomCropAndResize):
```
TensorArray dtype is float16 but Op is trying to write dtype float32 
```

I'd suspect this is the root cause and I'd check out what's happening there.

jaygala223 commented:
Hi @ianstenbit, sorry for the delay. I fixed a few of the tests and added flags for some. Still not able to get around the ones we discussed last time... I'll keep looking

Also, can you take a look at my fix for this layer (in pic) in the commit above? It seems hacky but does the job. We now stand at 6F, 21P, and 6S

![image](https://github.com/keras-team/keras-cv/assets/57001778/cf9ce230-7b85-416a-924c-54fda434480d)

ianstenbit commented on keras_cv/layers/preprocessing/with_mixed_precision_test.py:
```diff
@@ -155,16 +185,56 @@ def test_can_run_in_mixed_precision(self, layer_cls, init_args):
+        
         layer = layer_cls(**init_args)
+
+        if layer_cls == layers.CutMix:
```
CutMix doesn't support bounding boxes, so this should just be included in `NO_BOUNDING_BOX_TESTS`

jaygala223 replied:
Oh right!

jaygala223 commented:
Hi @ianstenbit, thanks for your tip. All tests are passing now!

> > Also, one of the tests is now showing this. A StackOverflow [thread](https://stackoverflow.com/questions/51248813/any-ideas-on-what-might-be-causing-this-tensorflow-error-object-was-never-used) suggests that it may be an internal bug with tf v1.2 but I'm on 2.12🤔
> 
> This is indeed a very weird error message. Looking at the error logs, it seems that there's a dtype issue in the same test (RandomCropAndResize):
> 
> ```
> TensorArray dtype is float16 but Op is trying to write dtype float32 
> ```
> 
> I'd suspect this is the root cause and I'd check out what's happening there.

ianstenbit commented on keras_cv/layers/preprocessing/with_mixed_precision_test.py:
```diff
@@ -30,6 +30,8 @@
             "crop_area_factor": (0.8, 1.0),
             "aspect_ratio_factor": (3 / 4, 4 / 3),
+            "bounding_box_format": "xywh",
+            "dtype": tf.float32,
```
Now that we're using Keras Core, we need to switch to string dtypes instead of `tf.float32`

So for the places in this config where we specify `"dtype": tf.float32` we should instead now specify `"dtype": "float32"`.

After that change I think we can merge this

jaygala223 replied:
Got it!

ianstenbit reviewed (changes requested):
Looks good to me -- just one small change to make sure CI is happy and then we can merge 😄

ianstenbit commented:
/gcbrun

ianstenbit commented on keras_cv/layers/preprocessing/with_mixed_precision_test.py:
```diff
@@ -156,14 +211,43 @@ def test_can_run_in_mixed_precision(self, layer_cls, init_args):
                     "operations `tf.image.adjust_saturation`, and "
                     "`tf.image.adjust_hue`. Skipping."
                 )
+            if layer_cls in NO_BOUNDING_BOXES_TESTS:
```
Our GPU testing caught this little bug for us -- this only gets checked if there's no GPU present because this is over-indented by one level.

But this also makes the NO_BOUNDING_BOX tests no-ops anyway. How about instead, we check if layer_cls is in `NO_BOUNDING_BOX_TESTS` and if it is we exclude the bounding boxes from the inputs to the layer instead? (So basically just an `if` statement down around line 250

jaygala223 replied:
Yes, that makes sense!

jaygala223 replied:
CI is still unhappy 😢 I made the change at line 250 as you suggested.

ianstenbit replied:
This line should not be here I think -- we can/should run these tests regardless of whether we have a GPU.

ianstenbit replied:
Unresolving this -- we should get rid of this check on line 214, as these tests should still be run in this case.

jaygala223 replied:
Oh okay

jaygala223 replied:
Hello, @ianstenbit, I removed the check on line 214 as you suggested. Sorry for the infinite little errors towards the end😢

jaygala223 commented:
All tests should be happy now 🥳

ianstenbit commented on keras_cv/layers/preprocessing/with_mixed_precision_test.py:
```diff
@@ -86,20 +97,54 @@
+            "dtype": "float32",
+        },
+    ),
+    # ("RandomShear", layers.RandomShear, {"x_factor": 0.3, "x_factor": 0.3}),
```
Let's either un-comment this or re-add it. Not sure when this originally got commented-out

jaygala223 replied:
This is weird... the test above it is the same one. I think I may have added it when fixing some styling issues. I'll just remove it to avoid duplicates.

ianstenbit commented:
Thanks for fixing the missing tests -- now one of them is failing. That's good because it means we're actually testing it now :-)

This should be the last bit:
```
FAILED keras_cv/layers/preprocessing/with_mixed_precision_test.py::WithMixedPrecisionTest::test_can_run_in_mixed_precision_CutMix - tensorflow.python.framework.errors_impl.InvalidArgumentError: Exception encountered when calling CutMix.call().

cannot compute Mul as input #1(zero-based) was expected to be a half tensor but is a float tensor [Op:Mul] name: 
```
Same issue as before, I expect that this is just because of a missing cast, or an extra one that needs to be added

jaygala223 commented:
I have fixed it. Thanks @ianstenbit, I couldn't have done it without your support and guidance. 🙌

ianstenbit commented:
/gcbrun

ianstenbit reviewed (approved):
Wohooo -- thank you!
